### DISCUSSION TITLE
CP-SAT Job Shop Problem: How to separate the tasks into SHIFTS? 1 worker/shift/day arrangement in flexible job shop.

### DISCUSSION THREAD

**Discussion by replicantdeca** [google/or-tools#1453] [CP-SAT questions]
Hello!

Would just like to separate tasks into **shifts** and make the worker assignment for each task adjust if the task is cut into shifts. 1 WORKER/SHIFT/DAY. Where should I modify this? Not sure if it is in the data part of the code...

Attached is my code with sample data. I am making the use of flexible job shop model
[flexible_jobshop.txt](https://github.com/google/or-tools/files/3435338/flexible_jobshop.txt)

**Comment by gregy4** [comment]
I have an idea ... optional interval with start and end variables in a shift domain, repeated per number of shifts and workers. In case optional interval is active, main interval variables (start and end) should be equal to variables in optional interval. Only one optional interval per main interval should be active. This way it should be not possible to have main interval over two or more shifts.

**Comment by replicantdeca** [comment]
Deeply sorry but I am a noob to CP and Python in general. Is it OK to ask for some form of skeletal code I can base off of?

**Comment by replicantdeca** [comment]
As seen in the .txt file with the code and sample data, the constraints implemented are the ff:
- Type 1 Machine can do Type 1 task, and so on. There are 4 types of machines for 4 types of tasks.
- Same goes with Workers.

I wish to implement this new constraint:
- Type 1 worker is assigned to Type 1 task assigned to Type 1 machine. However, at the current state, the tasks are not separated into shifts. The solver should separate tasks per shift (SHIFTS: 6am-2pm, 2pm-10pm, 10pm-6am) and assign different workers per shift. So, it will not just be Worker A assigned to Task X if it takes up 2 shifts. Task X is assigned to Worker A and Worker B who will work Shift 1 and Shift 2 respectively.

Appreciate all the help I could get! :(

**Comment by gregy4** [comment]
In case you want divide task between shifts and use different worker in each shift, you should probably model it as two or more intervals (in case you can choose from more workers optional intervals) based on maximum number of shifts that a task can be within at same time. All intervals should have variable duration. In case a task is in one shift only, duration of others should be zero. In other cases sum of duration over intervals should be equal to task duration.

Since it is not possible to create own propagator in cp-sat, you need for every possible position of a task say duration of these intervals. 

By the way many questions about shifts were answered in other issues or in google groups recently.

**Comment by replicantdeca** [comment]
Can I request a sort of skeletal code I can base off of? I am really illiterate in Python and need all the help I can get :/

Below is my code for the main interval as per flexible job shop code

```
suffix_name = '_j%i_t%i' % (job_id, task_id)
            start = model.NewIntVar(0, horizon, 'start' + suffix_name)
            duration = model.NewIntVar(min_duration, max_duration,
                                       'duration' + suffix_name)
            end = model.NewIntVar(0, horizon, 'end' + suffix_name)
            interval = model.NewIntervalVar(start, duration, end,
                                            'interval' + suffix_name)
```

**Comment by lperron** [comment]
I am closing modeling issues. Please repost on the mailing list if it is still relevent.
